File: impl/java/src/test/java/com/cptnfizzbin/keycard/FixturesTest.java
package com.cptnfizzbin.keycard;

import org.junit.Before;
import org.junit.Test;
import org.yaml.snakeyaml.Yaml;

import com.cptnfizzbin.keycard.policy.Policy;
import com.cptnfizzbin.keycard.policy.PolicyDefinition;
import com.cptnfizzbin.keycard.subject.SubjectDef;
import com.cptnfizzbin.keycard.subject.SubjectRef;
import com.cptnfizzbin.keycard.action.Action;
import com.cptnfizzbin.keycard.action.ActionFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Map;

import static org.junit.Assert.*;

public class FixturesTest {

    private static final String FIXTURES_DIR = "../../test/fixtures";
    private Yaml yaml = new Yaml();
    private Map<String, Object> objectsArticles;
    private Map<String, Object> objectsComments;
    private Map<String, Object> objectsUsers;

    @Before
    public void setUp() throws IOException {
        objectsArticles = loadObjectsYaml("articles.yaml");
        objectsComments = loadObjectsYaml("comments.yaml");
        objectsUsers = loadObjectsYaml("users.yaml");
    }

    private String loadPolicyYaml(String filename) throws IOException {
        String path = FIXTURES_DIR + "/policies/" + filename;
        return new String(Files.readAllBytes(Paths.get(path)));
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> loadObjectsYaml(String filename) throws IOException {
        String path = FIXTURES_DIR + "/objects/" + filename;
        String content = new String(Files.readAllBytes(Paths.get(path)));
        Map<String, Object> loaded = yaml.load(content);
        
        // Extract the object collection key (e.g., "articles" from articles.yaml)
        // The YAML file has structure: articles: { article_1: {...}, ... }
        // We need to return just the inner map { article_1: {...}, ... }
        for (Object value : loaded.values()) {
            if (value instanceof Map) {
                return (Map<String, Object>) value;
            }
        }
        return loaded;
    }

    @SuppressWarnings("unchecked")
    private PolicyDefinition loadPolicyDef(String filename) throws IOException {
        String content = loadPolicyYaml(filename);
        Map<String, Object> parsed = yaml.load(content);
        
        int version = (int) parsed.getOrDefault("version", 1);
        java.util.List<PolicyDefinition.Rule> allowRules = new java.util.ArrayList<>();
        java.util.List<PolicyDefinition.Rule> denyRules = new java.util.ArrayList<>();
        
        java.util.List<?> allows = (java.util.List<?>) parsed.getOrDefault("allow", java.util.List.of());
        for (Object rule : allows) {
            if (rule instanceof java.util.List) {
                allowRules.add(convertArrayToRule((java.util.List<?>) rule));
            }
        }
        
        java.util.List<?> denies = (java.util.List<?>) parsed.getOrDefault("deny", java.util.List.of());
        for (Object rule : denies) {
            if (rule instanceof java.util.List) {
                denyRules.add(convertArrayToRule((java.util.List<?>) rule));
            }
        }
        
        return new PolicyDefinition(version, allowRules, denyRules);
    }

    @SuppressWarnings("unchecked")
    private Policy createPolicy(String policyFile) throws IOException {
        String content = loadPolicyYaml(policyFile);
        Map<String, Object> parsed = yaml.load(content);
        
        int version = (int) parsed.getOrDefault("version", 1);
        java.util.List<PolicyDefinition.Rule> allowRules = new java.util.ArrayList<>();
        java.util.List<PolicyDefinition.Rule> denyRules = new java.util.ArrayList<>();
        
        java.util.List<?> allows = (java.util.List<?>) parsed.getOrDefault("allow", java.util.List.of());
        for (Object rule : allows) {
            if (rule instanceof java.util.List) {
                allowRules.add(convertArrayToRule((java.util.List<?>) rule));
            }
        }
        
        java.util.List<?> denies = (java.util.List<?>) parsed.getOrDefault("deny", java.util.List.of());
        for (Object rule : denies) {
            if (rule instanceof java.util.List) {
                denyRules.add(convertArrayToRule((java.util.List<?>) rule));
            }
        }
        
        return new Policy(new PolicyDefinition(version, allowRules, denyRules));
    }

    @SuppressWarnings("unchecked")
    private PolicyDefinition.Rule convertArrayToRule(java.util.List<?> ruleArray) {
        String action = (String) ruleArray.get(0);
        String subject = (String) ruleArray.get(1);
        Map<String, Object> conditions = null;
        
        if (ruleArray.size() > 2) {
            conditions = (Map<String, Object>) ruleArray.get(2);
        }
        
        return new PolicyDefinition.Rule(action, subject, conditions);
    }

    private <T> SubjectRef<T> wrapSubject(String subjectName, T obj) {
        SubjectDef<T> def = SubjectDef.create(subjectName, (Class<T>) obj.getClass());
        return def.wrap(obj);
    }

    private Object addSubjectType(Map<String, Object> obj, String typeName) {
        Map<String, Object> result = new java.util.HashMap<>(obj);
        result.put("__name", typeName);
        return result;
    }

    // Policy 1: Simple
    @Test
    public void testPolicy01DataLoadsCorrectly() throws IOException {
        assertNotNull("Articles should be loaded", objectsArticles);
        assertNotNull("Article 1 should be present", objectsArticles.get("article_1"));
        @SuppressWarnings("unchecked")
        Map<String, Object> article1 = (Map<String, Object>) objectsArticles.get("article_1");
        assertEquals("Article 1 owner_id should be 1", 1, article1.get("owner_id"));
    }

    @Test
    public void testPolicy01AllowsRead() throws IOException {
        Policy policy = createPolicy("policy-01-simple.yaml");
        assertTrue("Policy should allow Read on Article", policy.can("Read", "Article"));
    }

    @Test
    public void testPolicy01AllowsCreateForOwner1() throws IOException {
        Policy policy = createPolicy("policy-01-simple.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_1");
        Action<?> createAction = ActionFactory.create("Create");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Create for owner_id 1", policy.can(createAction, subject));
    }

    @Test
    public void testPolicy01DeniesCreateForOwner2() throws IOException {
        Policy policy = createPolicy("policy-01-simple.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_3");
        Action<?> createAction = ActionFactory.create("Create");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertFalse("Policy should deny Create for owner_id 2", policy.can(createAction, subject));
    }

    // Policy 2: Ownership
    @Test
    public void testPolicy02AllowsRead() throws IOException {
        Policy policy = createPolicy("policy-02-ownership.yaml");
        assertTrue("Policy should allow Read on Article", policy.can("Read", "Article"));
    }

    @Test
    public void testPolicy02AllowsUpdateForOwner() throws IOException {
        Policy policy = createPolicy("policy-02-ownership.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_1");
        Action<?> updateAction = ActionFactory.create("Update");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Update for owner", policy.can(updateAction, subject));
    }

    @Test
    public void testPolicy02DeniesUpdateForNonOwner() throws IOException {
        Policy policy = createPolicy("policy-02-ownership.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_3");
        Action<?> updateAction = ActionFactory.create("Update");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertFalse("Policy should deny Update for non-owner", policy.can(updateAction, subject));
    }

    @Test
    public void testPolicy02AllowsDeleteForOwner() throws IOException {
        Policy policy = createPolicy("policy-02-ownership.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_1");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Delete for owner", policy.can(deleteAction, subject));
    }

    @Test
    public void testPolicy02DeniesDeleteForNonOwner() throws IOException {
        Policy policy = createPolicy("policy-02-ownership.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_3");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertFalse("Policy should deny Delete for non-owner", policy.can(deleteAction, subject));
    }

    // Policy 3: Complex Conditions
    @Test
    public void testPolicy03AllowsRead() throws IOException {
        Policy policy = createPolicy("policy-03-complex-conditions.yaml");
        assertTrue("Policy should allow Read on Article", policy.can("Read", "Article"));
    }

    @Test
    public void testPolicy03AllowsUpdateForOwnerNonArchived() throws IOException {
        Policy policy = createPolicy("policy-03-complex-conditions.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_1");
        Action<?> updateAction = ActionFactory.create("Update");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Update for owner with non-archived status", policy.can(updateAction, subject));
    }

    @Test
    public void testPolicy03DeniesUpdateForArchived() throws IOException {
        Policy policy = createPolicy("policy-03-complex-conditions.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_3");
        Action<?> updateAction = ActionFactory.create("Update");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertFalse("Policy should deny Update for archived articles", policy.can(updateAction, subject));
    }

    @Test
    public void testPolicy03DeniesDeleteArchived() throws IOException {
        Policy policy = createPolicy("policy-03-complex-conditions.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_3");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertFalse("Policy should deny Delete on archived articles", policy.can(deleteAction, subject));
    }

    @Test
    public void testPolicy03AllowsDeleteNonArchived() throws IOException {
        Policy policy = createPolicy("policy-03-complex-conditions.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_1");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Delete on non-archived articles", policy.can(deleteAction, subject));
    }

    // Policy 4: Multi-Resource
    @Test
    public void testPolicy04AllowsReadArticle() throws IOException {
        Policy policy = createPolicy("policy-04-multi-resource.yaml");
        assertTrue("Policy should allow Read on Article", policy.can("Read", "Article"));
    }

    @Test
    public void testPolicy04AllowsReadComment() throws IOException {
        Policy policy = createPolicy("policy-04-multi-resource.yaml");
        assertTrue("Policy should allow Read on Comment", policy.can("Read", "Comment"));
    }

    @Test
    public void testPolicy04AllowsReadUser() throws IOException {
        Policy policy = createPolicy("policy-04-multi-resource.yaml");
        assertTrue("Policy should allow Read on User", policy.can("Read", "User"));
    }

    @Test
    public void testPolicy04AllowsUpdateCommentForAuthor() throws IOException {
        Policy policy = createPolicy("policy-04-multi-resource.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> comment = (Map<String, Object>) objectsComments.get("comment_1");
        Action<?> updateAction = ActionFactory.create("Update");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Comment", comment);
        
        assertTrue("Policy should allow Update Comment for author", policy.can(updateAction, subject));
    }

    @Test
    public void testPolicy04DeniesUpdateCommentForNonAuthor() throws IOException {
        Policy policy = createPolicy("policy-04-multi-resource.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> comment = (Map<String, Object>) objectsComments.get("comment_4");
        Action<?> updateAction = ActionFactory.create("Update");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Comment", comment);
        
        assertFalse("Policy should deny Update Comment for non-author", policy.can(updateAction, subject));
    }

    @Test
    public void testPolicy04DeniesDeleteUser() throws IOException {
        Policy policy = createPolicy("policy-04-multi-resource.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) objectsUsers.get("user_1");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("User", user);
        
        assertFalse("Policy should deny Delete User entirely", policy.can(deleteAction, subject));
    }

    @Test
    public void testPolicy04DeniesDeleteAnyUser() throws IOException {
        Policy policy = createPolicy("policy-04-multi-resource.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) objectsUsers.get("user_3");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("User", user);
        
        assertFalse("Policy should deny Delete any user", policy.can(deleteAction, subject));
    }

    // Policy 5: Advanced
    @Test
    public void testPolicy05AllowsRead() throws IOException {
        Policy policy = createPolicy("policy-05-advanced.yaml");
        assertTrue("Policy should allow Read on Article", policy.can("Read", "Article"));
    }

    @Test
    public void testPolicy05AllowsCreate() throws IOException {
        Policy policy = createPolicy("policy-05-advanced.yaml");
        assertTrue("Policy should allow Create on Article", policy.can("Create", "Article"));
    }

    @Test
    public void testPolicy05AllowsUpdateForOwnerDraft() throws IOException {
        Policy policy = createPolicy("policy-05-advanced.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_2");
        Action<?> updateAction = ActionFactory.create("Update");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Update for owner with draft/published status", policy.can(updateAction, subject));
    }

    @Test
    public void testPolicy05AllowsDeleteLowViewCount() throws IOException {
        Policy policy = createPolicy("policy-05-advanced.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_4");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Delete for owner with low view count", policy.can(deleteAction, subject));
    }

    @Test
    public void testPolicy05DeniesDeleteHighViewArchived() throws IOException {
        Policy policy = createPolicy("policy-05-advanced.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_3");
        Action<?> deleteAction = ActionFactory.create("Delete");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertFalse("Policy should deny Delete for published high-view articles", policy.can(deleteAction, subject));
    }

    @Test
    public void testPolicy05AllowsPublishForOwnerValidTitle() throws IOException {
        Policy policy = createPolicy("policy-05-advanced.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_2");
        Action<?> publishAction = ActionFactory.create("Publish");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertTrue("Policy should allow Publish for owner with valid title", policy.can(publishAction, subject));
    }

    @Test
    public void testPolicy05DeniesPublishForNonOwner() throws IOException {
        Policy policy = createPolicy("policy-05-advanced.yaml");
        @SuppressWarnings("unchecked")
        Map<String, Object> article = (Map<String, Object>) objectsArticles.get("article_5");
        Action<?> publishAction = ActionFactory.create("Publish");
        SubjectRef<Map<String, Object>> subject = wrapSubject("Article", article);
        
        assertFalse("Policy should deny Publish for non-owner", policy.can(publishAction, subject));
    }
}

File: impl/node/src/fixtures.test.ts
import * as fs from "fs";
import * as path from "path";
import * as YAML from "yaml";
import { describe, test, beforeAll, expect } from "vitest";
import { Policy } from "./index";

describe("Fixture-based Policy Tests", () => {
  const fixturesDir = path.join(__dirname, "../../../test/fixtures");

  function loadPolicyDef(filename: string): any {
    const filePath = path.join(fixturesDir, "policies", filename);
    const content = fs.readFileSync(filePath, "utf-8");
    const parsed = YAML.parse(content);
    // Transform YAML format to PolicyDefinition format
    return {
      version: parsed.version,
      rules: {
        allow: parsed.allow || [],
        deny: parsed.deny || [],
      },
    };
  }

  function loadObjects(): any {
    const articlesPath = path.join(fixturesDir, "objects", "articles.yaml");
    const commentsPath = path.join(fixturesDir, "objects", "comments.yaml");
    const usersPath = path.join(fixturesDir, "objects", "users.yaml");

    const result: any = {};

    if (fs.existsSync(articlesPath)) {
      const articlesData = YAML.parse(fs.readFileSync(articlesPath, "utf-8")).articles;
      result.articles = {};
      for (const [key, value] of Object.entries(articlesData)) {
        // Add __name property to identify subject type for policy checking
        result.articles[key] = { ...value as any, __name: "Article" };
      }
    }
    if (fs.existsSync(commentsPath)) {
      const commentsData = YAML.parse(fs.readFileSync(commentsPath, "utf-8")).comments;
      result.comments = {};
      for (const [key, value] of Object.entries(commentsData)) {
        result.comments[key] = { ...value as any, __name: "Comment" };
      }
    }
    if (fs.existsSync(usersPath)) {
      const usersData = YAML.parse(fs.readFileSync(usersPath, "utf-8")).users;
      result.users = {};
      for (const [key, value] of Object.entries(usersData)) {
        result.users[key] = { ...value as any, __name: "User" };
      }
    }

    return result;
  }

  describe("Policy 1: Simple", () => {
    let policy: Policy;
    let objects: any;

    beforeAll(() => {
      const policyDef = loadPolicyDef("policy-01-simple.yaml");
      policy = new Policy(policyDef);
      objects = loadObjects();
    });

    test("data loads correctly", () => {
      expect(objects.articles).toBeDefined();
      expect(objects.articles.article_1).toBeDefined();
      expect(objects.articles.article_1.owner_id).toBe(1);
    });

    test("allows Read on Article", () => {
      expect(policy.can("Read" as any, "Article" as any)).toBe(true);
    });

    test("allows Create on Article for owner_id 1", () => {
      const article = objects.articles.article_1;
      expect(policy.can("Create" as any, article)).toBe(true);
    });

    test("denies Create for owner_id 2", () => {
      const article = objects.articles.article_3;
      expect(policy.can("Create" as any, article)).toBe(false);
    });
  });

  describe("Policy 2: Ownership", () => {
    let policy: Policy;
    let objects: any;

    beforeAll(() => {
      const policyDef = loadPolicyDef("policy-02-ownership.yaml");
      policy = new Policy(policyDef);
      objects = loadObjects();
    });

    test("allows Read on Article", () => {
      expect(policy.can("Read" as any, "Article" as any)).toBe(true);
    });

    test("allows Update for owner", () => {
      const article = objects.articles.article_1;
      expect(policy.can("Update" as any, article)).toBe(true);
    });

    test("denies Update for non-owner", () => {
      const article = objects.articles.article_3;
      expect(policy.can("Update" as any, article)).toBe(false);
    });

    test("allows Delete for owner", () => {
      const article = objects.articles.article_1;
      expect(policy.can("Delete" as any, article)).toBe(true);
    });

    test("denies Delete for non-owner", () => {
      const article = objects.articles.article_3;
      expect(policy.can("Delete" as any, article)).toBe(false);
    });
  });

  describe("Policy 3: Complex Conditions", () => {
    let policy: Policy;
    let objects: any;

    beforeAll(() => {
      const policyDef = loadPolicyDef("policy-03-complex-conditions.yaml");
      policy = new Policy(policyDef);
      objects = loadObjects();
    });

    test("allows Read on Article", () => {
      expect(policy.can("Read" as any, "Article" as any)).toBe(true);
    });

    test("allows Update for owner with non-archived status", () => {
      const article = objects.articles.article_1;
      expect(policy.can("Update" as any, article)).toBe(true);
    });

    test("denies Update for archived articles", () => {
      const article = objects.articles.article_3;
      expect(policy.can("Update" as any, article)).toBe(false);
    });

    test("denies Delete on archived articles", () => {
      const article = objects.articles.article_3;
      expect(policy.can("Delete" as any, article)).toBe(false);
    });

    test("allows Delete on non-archived articles", () => {
      const article = objects.articles.article_1;
      expect(policy.can("Delete" as any, article)).toBe(true);
    });
  });

  describe("Policy 4: Multi-Resource", () => {
    let policy: Policy;
    let objects: any;

    beforeAll(() => {
      const policyDef = loadPolicyDef("policy-04-multi-resource.yaml");
      policy = new Policy(policyDef);
      objects = loadObjects();
    });

    test("allows Read on Article", () => {
      expect(policy.can("Read" as any, "Article" as any)).toBe(true);
    });

    test("allows Read on Comment", () => {
      expect(policy.can("Read" as any, "Comment" as any)).toBe(true);
    });

    test("allows Read on User", () => {
      expect(policy.can("Read" as any, "User" as any)).toBe(true);
    });

    test("allows Update Comment for author", () => {
      const comment = objects.comments.comment_1;
      expect(policy.can("Update" as any, comment)).toBe(true);
    });

    test("denies Update Comment for non-author", () => {
      const comment = objects.comments.comment_4;
      expect(policy.can("Update" as any, comment)).toBe(false);
    });

    test("denies Delete User entirely", () => {
      const user = objects.users.user_1;
      expect(policy.can("Delete" as any, user)).toBe(false);
    });

    test("denies Delete any user", () => {
      const user = objects.users.user_3;
      expect(policy.can("Delete" as any, user)).toBe(false);
    });
  });

  describe("Policy 5: Advanced", () => {
    let policy: Policy;
    let objects: any;

    beforeAll(() => {
      const policyDef = loadPolicyDef("policy-05-advanced.yaml");
      policy = new Policy(policyDef);
      objects = loadObjects();
    });

    test("allows Read on Article", () => {
      expect(policy.can("Read" as any, "Article" as any)).toBe(true);
    });

    test("allows Create on Article", () => {
      expect(policy.can("Create" as any, "Article" as any)).toBe(true);
    });

    test("allows Update for owner with draft/published status", () => {
      const article = objects.articles.article_2;
      expect(article.owner_id).toBe(1);
      expect(article.status).toBe("draft");
      expect(policy.can("Update" as any, article)).toBe(true);
    });

    test("allows Delete for owner with low view count", () => {
      const article = objects.articles.article_4;
      expect(article.owner_id).toBe(1);
      expect(article.view_count).toBe(50);
      expect(policy.can("Delete" as any, article)).toBe(true);
    });

    test("denies Delete for published high-view articles", () => {
      const article = objects.articles.article_3;
      expect(article.view_count).toBe(2000);
      expect(article.status).toBe("archived");
      expect(policy.can("Delete" as any, article)).toBe(false);
    });

    test("allows Publish for owner with valid title", () => {
      const article = objects.articles.article_2;
      expect(article.owner_id).toBe(1);
      expect(article.status).toBe("draft");
      expect(article.title).toMatch(/^[A-Z]/);
      expect(policy.can("Publish" as any, article)).toBe(true);
    });

    test("denies Publish for non-owner", () => {
      const article = objects.articles.article_5;
      expect(article.owner_id).toBe(3);
      expect(policy.can("Publish" as any, article)).toBe(false);
    });
  });
});
